### PR TITLE
Feature/likes-rest

### PR DESCRIPTION
**Pull request by NikPaukov**

### CODE REVIEW

File: src/services/likes.service.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  createOne,
+  getOne,
+  deleteOne,
+  updateOne,
+} from "../repositories/likes.repository";
+import { getUserById } from "../repositories/user.repository";
+import { getTweetById, incrementTweetStats } from "./tweets.service";
+
+export const changeLike = async (tweetId: string, userId: string) => {
+  if (getUserById(userId) && getTweetById(tweetId)) {
```
Comment (alexandrtovmach) [inline src/services/likes.service.ts]: not sure why do you need this checks, but anyway they should be async

File: src/services/likes.service.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    const exists = !!(await getOne({ userId, tweetId }));
+    if (!exists) {
+      await incrementTweetStats(tweetId, "stats.likes", 1);
+      return createOne(tweetId, userId);
+    } else {
+      await incrementTweetStats(tweetId, "stats.likes", -1);
+      return deleteOne({ tweetId, userId });
```
Comment (alexandrtovmach) [inline src/services/likes.service.ts]: I'd recommend make it a bit smarter, to reduce count of requessts:
1. try to delete like with userId and tweetId
2. if it failed, it means, nothing to delete, then create a new one and increase stats
3. if it success, decrease stats

and you potentially save time of one request per each operation

**Review by alexandrtovmach — CHANGES_REQUESTED** [review]

**Comment by alexandrtovmach** [comment]
and one more thing, let's use normal name for references: "instance" instead of "instanceId"

**Comment by alexandrtovmach** [comment]
you have merge conflict artifacts in your PR

File: src/models/Like.model.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import { ObjectId } from "mongodb";
+import { model, Schema, Document } from "mongoose";
+
+export interface Like extends Document {
+  userId: ObjectId;
+  tweetId: ObjectId;
```
Comment (VitaliiDenysiuk) [inline src/models/Like.model.ts]:
```suggestion
  user: ObjectId;
  tweet: ObjectId;
```
@alexandrtovmach ordered delete "Id" at all models

**Review by VitaliiDenysiuk — CHANGES_REQUESTED** [review]

File: README.md [hunk 4]
```diff
@@ -16,10 +16,24 @@
...
+- Registration - post: 127.0.0.1:3001/api/auth/register
+- Log in - post: 127.0.0.1:3001/api/auth/login
+- Refresh access token - post: 127.0.0.1:3001/api/auth/refresh
+- test authorized - get: 127.0.0.1:3001/api/testAuthorized
+- Proxy Link Preview - `http://127.0.0.1:3001/api/link-preview/?url=${url}`
+- Pagination query info: `https://www.npmjs.com/package/mongoose-paginate-v2`, Url example: `127.0.0.1:3001/api/tweets/?limit=1&page=2&query={"content": "bbb"}`
+- Likes: Post `127.0.0.1:3001/api/likes` with twitId & userId - use for likes and dislikes
+  how it works Frontend:
+  =======
+
```
Comment (VitaliiDenysiuk) [inline README.md]: please, check it correct. I made some order in that file

File: src/models/Like.model.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+}
+
+const LikeSchema: Schema = new Schema<Like>({
+  userId: {
+    type: ObjectId,
+    required: true,
+    ref: "User",
```
Comment (VitaliiDenysiuk) [inline src/models/Like.model.ts]:
```suggestion
    ref: "Users",
```

File: src/models/Like.model.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  },
+  tweetId: {
```
Comment (VitaliiDenysiuk) [inline src/models/Like.model.ts]:
```suggestion
  tweet: {
```

File: src/models/Like.model.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import { ObjectId } from "mongodb";
+import { model, Schema, Document } from "mongoose";
+
+export interface Like extends Document {
+  userId: ObjectId;
+  tweetId: ObjectId;
+}
+
+const LikeSchema: Schema = new Schema<Like>({
+  userId: {
```
Comment (VitaliiDenysiuk) [inline src/models/Like.model.ts]:
```suggestion
  user: {
```

File: src/repositories/likes.repository.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
+import { LikeModel } from "../models/Like.model";
+
+export const getOne = async (
+  data: { _id: string } | { userId: string; tweetId: string }
```
Comment (VitaliiDenysiuk) [inline src/repositories/likes.repository.ts]:
```suggestion
  data: { _id: string } | { user: string; tweet: string }
```

File: src/repositories/likes.repository.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  return LikeModel.findOne(data);
+};
+export const deleteOne = async (
+  data:
+    | {
+        _id: string;
+      }
+    | {
+        userId: string;
+        tweetId: string;
```
Comment (VitaliiDenysiuk) [inline src/repositories/likes.repository.ts]:
```suggestion
        user: string;
        tweet: string;
```

File: src/repositories/likes.repository.ts [hunk 10]
```diff
@@ -0,0 +1,32 @@
+import { LikeModel } from "../models/Like.model";
+
+export const getOne = async (
+  data: { _id: string } | { userId: string; tweetId: string }
+) => {
+  return LikeModel.findOne(data);
+};
+export const deleteOne = async (
```
Comment (VitaliiDenysiuk) [inline src/repositories/likes.repository.ts]:
```suggestion
export const deleteOne =  (
```
no await in that function

File: src/repositories/likes.repository.ts [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+        _id: string;
+      }
+    | {
+        userId: string;
+        tweetId: string;
+      }
+) => {
+  return LikeModel.deleteOne(data);
+};
+export const createOne = async (tweetId: string, userId: string) => {
```
Comment (VitaliiDenysiuk) [inline src/repositories/likes.repository.ts]:
```suggestion
export const createOne =  (tweetId: string, userId: string) => {
```

File: src/repositories/likes.repository.ts [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+  const likeModel = new LikeModel({
+    userId,
+    tweetId,
```
Comment (VitaliiDenysiuk) [inline src/repositories/likes.repository.ts]:
```suggestion
    user,
    tweet,
```

File: src/repositories/likes.repository.ts [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+  });
+  return likeModel.save();
+};
+export const updateOne = async (
```
Comment (VitaliiDenysiuk) [inline src/repositories/likes.repository.ts]:
```suggestion
export const updateOne =  (
```
no await in that function

File: src/repositories/likes.repository.ts [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+  id: string,
+  newData: { tweetId?: string; userId?: string }
```
Comment (VitaliiDenysiuk) [inline src/repositories/likes.repository.ts]:
```suggestion
  newData: { tweet?: string; user?: string }
```

File: src/repositories/tweet.repository.ts [hunk 15]
```diff
@@ -28,8 +28,20 @@ export const updateOne = async (
     content?: string;
     authorId?: string;
```
Comment (VitaliiDenysiuk) [inline src/repositories/tweet.repository.ts]: bad merge no such field

File: src/repositories/tweet.repository.ts [hunk 16]
```diff
@@ -28,8 +28,20 @@ export const updateOne = async (
...
 ) => {
-  return TweetModel.updateOne({ _id: id }, newData);
+  return TweetModel.findOneAndUpdate({ _id: id }, newData);
+};
+export const changeStats = async (
+  id: string,
+  fieldName: string,
+  value: number
+) => {
+  const res = await TweetModel.updateOne(
```
Comment (VitaliiDenysiuk) [inline src/repositories/tweet.repository.ts]: Why did you create const? 
Return result strait

**Review by VitaliiDenysiuk — CHANGES_REQUESTED** [review]

File: src/routes/likes.route.ts [hunk 17]
```diff
@@ -0,0 +1,24 @@
+import express, { Request, Response } from "express";
+
+import { changeLike, getLikeById, updateLike } from "../services/likes.service";
+
+const likesRouter = express.Router();
+
+likesRouter.post("/", async (req: Request, res: Response) => {
+  await changeLike(req.body.tweetId, req.body.userId);
```
Comment (VitaliiDenysiuk) [inline src/routes/likes.route.ts]:
```suggestion
  await changeLike(req.body.tweet, req.body.user);
```

File: src/routes/likes.route.ts [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+});
+
+likesRouter.get("/:id", async (req: Request, res: Response) => {
+  const result = await getLikeById(req.params.id);
+  res.status(200).send(result);
+});
+likesRouter.put("/:id", async (req: Request, res: Response) => {
+  const result = await updateLike(req.params.id, {
+    tweetId: req.body.tweetId,
+    userId: req.body.userId,
```
Comment (VitaliiDenysiuk) [inline src/routes/likes.route.ts]:
```suggestion
    tweetId: req.body.tweet,
    userId: req.body.user,
```

File: src/services/likes.service.ts [hunk 19]
```diff
@@ -0,0 +1,27 @@
+import {
+  createOne,
+  getOne,
+  deleteOne,
+  updateOne,
+} from "../repositories/likes.repository";
+import { changeTweetStats } from "./tweets.service";
+
+export const changeLike = async (tweetId: string, userId: string) => {
+  const deleted = await deleteOne({ userId, tweetId });
```
Comment (VitaliiDenysiuk) [inline src/services/likes.service.ts]:
```suggestion
  const deleted = await deleteOne({ user, tweet });
```

File: src/services/likes.service.ts [hunk 20]
```diff
@@ -0,0 +1,27 @@
+import {
+  createOne,
+  getOne,
+  deleteOne,
+  updateOne,
+} from "../repositories/likes.repository";
+import { changeTweetStats } from "./tweets.service";
+
+export const changeLike = async (tweetId: string, userId: string) => {
```
Comment (VitaliiDenysiuk) [inline src/services/likes.service.ts]:
```suggestion
export const changeLike = async (tweet: string, user: string) => {
```

File: src/services/likes.service.ts [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+  const deleted = await deleteOne({ userId, tweetId });
+  if (deleted.deletedCount) {
+    await changeTweetStats(tweetId, "stats.likes", -1);
+  } else {
+    await changeTweetStats(tweetId, "stats.likes", 1);
+    await createOne(tweetId, userId);
```
Comment (VitaliiDenysiuk) [inline src/services/likes.service.ts]:
```suggestion
    await changeTweetStats(tweet, "stats.likes", -1);
  } else {
    await changeTweetStats(tweet, "stats.likes", 1);
    await createOne(tweet, user);
```

File: src/services/likes.service.ts [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+  }
+};
+
+export const getLikeById = async (id: string) => {
+  return getOne({ _id: id });
+};
+export const updateLike = async (
+  id: string,
+  newData: { tweetId?: string; userId?: string }
```
Comment (VitaliiDenysiuk) [inline src/services/likes.service.ts]:
```suggestion
  newData: { tweet?: string; user?: string }
```

File: src/services/likes.service.ts [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+  const deleted = await deleteOne({ userId, tweetId });
+  if (deleted.deletedCount) {
+    await changeTweetStats(tweetId, "stats.likes", -1);
+  } else {
+    await changeTweetStats(tweetId, "stats.likes", 1);
+    await createOne(tweetId, userId);
+  }
+};
+
+export const getLikeById = async (id: string) => {
```
Comment (VitaliiDenysiuk) [inline src/services/likes.service.ts]:
```suggestion
export const getLikeById =  (id: string) => {
```
await absent

File: src/services/tweets.service.ts [hunk 24]
```diff
@@ -34,8 +35,16 @@ export const updateTweet = async (
     content?: string;
     authorId?: string;
```
Comment (VitaliiDenysiuk) [inline src/services/tweets.service.ts]:
```suggestion
    author?: string;
```
bad merge

**Review by VitaliiDenysiuk — CHANGES_REQUESTED** [review]

**Review by alexandrtovmach — APPROVED** [review]

**Pull request merged**